File: app/routers/integration_sync.py
"""Queue controls and scoped review/freshness information."""
import json
import time
from fastapi import APIRouter,Depends,HTTPException
from pydantic import BaseModel,Field
from ..database import db_dependency
from ..integration_sync import discover,enabled
from .extras import set_setting

router=APIRouter(prefix='/api/integration-sync',tags=['Integration sync'])

def admin(c):
    if 'admin' not in c.user['roles']:raise HTTPException(403,'Only administrators can manage integration synchronization')

def rows(c,cid=None):
    query='SELECT j.*,c.name AS clinic_name FROM integration_jobs j JOIN clinics c ON c.id=j.clinic_id'
    result=[]
    for r in c.execute(query+(' WHERE j.clinic_id=?' if cid else '')+' ORDER BY j.provider,j.next_at', (cid,) if cid else ()):
        d=dict(r);d.pop('lease');cfg=json.loads(d.pop('config'))
        d['site_name']=cfg.get('name') or 'Main Site'
        d['running']=r['lease_until']>time.time()
        d['pending']=c.execute("SELECT COUNT(*) FROM integration_changes WHERE job_id=? AND status='pending'",(r['id'],)).fetchone()[0]
        d['overrides']=c.execute('SELECT COUNT(*) FROM integration_fields WHERE job_id=? AND overridden=1',(r['id'],)).fetchone()[0]
        d['warnings']=json.loads(d['warnings']);result.append(d)
    return result

@router.get('/settings')
def settings(c=Depends(db_dependency)):
    admin(c)
    return {'enabled':enabled(c),'jobs':rows(c)}

class Settings(BaseModel): enabled:bool

@router.put('/settings')
def configure(p:Settings,c=Depends(db_dependency)):
    admin(c);previous=enabled(c);discover(c)
    set_setting(c,'integration_sync_enabled','1' if p.enabled else '0')
    if p.enabled and not previous:
        jobs=list(c.execute('SELECT id,interval_minutes FROM integration_jobs WHERE enabled=1 ORDER BY provider,id'))
        for index,r in enumerate(jobs):c.execute('UPDATE integration_jobs SET next_at=? WHERE id=?',(time.time()+index*r['interval_minutes']*60/max(1,len(jobs)),r['id']))
    return settings(c)

@router.post('/discover')
def discover_jobs(c=Depends(db_dependency)):
    admin(c);discover(c);return settings(c)

class JobSettings(BaseModel):
    enabled:bool
    interval_minutes:int=Field(default=30,ge=15,le=1440)

@router.put('/jobs/{jid}')
def job_settings(jid:int,p:JobSettings,c=Depends(db_dependency)):
    admin(c)
    if not c.execute('SELECT id FROM integration_jobs WHERE id=?',(jid,)).fetchone():raise HTTPException(404,'Sync job not found')
    c.execute('UPDATE integration_jobs SET enabled=?,interval_minutes=? WHERE id=?',(p.enabled,p.interval_minutes,jid))
    return {'saved':True}

@router.post('/jobs/{jid}/now')
def now(jid:int,c=Depends(db_dependency)):
    admin(c)
    if not enabled(c):raise HTTPException(409,'Enable background synchronization first; Sync now uses the shared queue')
    job=c.execute('SELECT * FROM integration_jobs WHERE id=?',(jid,)).fetchone()
    if not job:raise HTTPException(404,'Sync job not found')
    if not job['enabled']:raise HTTPException(409,'Resume this site first')
    if not job['lease_until']>time.time():c.execute('UPDATE integration_jobs SET next_at=? WHERE id=?',(time.time(),jid))
    return {'queued':True}

@router.get('/clinics/{cid}')
def clinic(cid:int,c=Depends(db_dependency)):
    if not c.execute('SELECT id FROM clinics WHERE id=?',(cid,)).fetchone():raise HTTPException(404,'Clinic not found')
    return {'enabled':enabled(c),'jobs':rows(c,cid),'changes':[dict(r)|{'detail':json.loads(r['detail'])} for r in c.execute("SELECT * FROM integration_changes WHERE clinic_id=? AND status='pending' ORDER BY updated_at DESC LIMIT 200",(cid,))]}

@router.post('/changes/{change_id}/acknowledge')
def acknowledge(change_id:int,c=Depends(db_dependency)):
    admin(c)
    if not c.execute('SELECT id FROM integration_changes WHERE id=?',(change_id,)).fetchone():raise HTTPException(404,'Change not found')
    c.execute("UPDATE integration_changes SET status='reviewed' WHERE id=?",(change_id,))
    return {'reviewed':True}

File: app/static/js/integration-sync.js
import {api} from './api.js';
import {esc,openModal,toast,navigate} from './ui.js';
const date=t=>t?new Date(t*1000).toLocaleString():'Not yet';
const state=(j,global)=>!global?'Globally paused':!j.enabled?'Paused':j.running?'Syncing':j.error?'Retry scheduled':j.next_at<Date.now()/1000?'Queued / delayed':'Scheduled';

export async function syncSettings(container){
  const card=document.createElement('section');card.className='card mb';
  (container.querySelector('section[aria-label="AI & integrations"]')||container).append(card);
  async function load(){
    const data=await api.post('/api/integration-sync/discover',{});
    if(!card.isConnected)return;
    card.innerHTML=`<h2>Automatic integration sync</h2><p>Read-only source refreshes, staggered between linked clinics. New machines and wiring changes need review. Manual overrides are preserved. The interval is a target; provider limits may delay a refresh.</p>
      <label><input type="checkbox" data-enabled ${data.enabled?'checked':''}> Enable background synchronization</label><button class="btn" data-save>Save global setting</button><button class="btn" data-reload>Refresh status</button><p role="status"></p>
      ${data.jobs.map(j=>`<details class="card"><summary>${esc(j.clinic_name)} · ${esc(j.provider)} · ${esc(j.site_name)} — ${esc(state(j,data.enabled))}</summary><p>Last success: ${date(j.last_success)}<br>Next attempt: ${date(j.next_at)}<br>${j.pending} changes to review · ${j.overrides} overridden fields</p><p>${esc(j.error||'')}${j.warnings.map(w=>'<br>'+esc(w)).join('')}</p>
        <label><input type="checkbox" data-active="${j.id}" ${j.enabled?'checked':''}> Sync this site</label><label>Interval (minutes)<input type="number" min="15" max="1440" value="${j.interval_minutes}" data-interval="${j.id}"></label><div class="actions"><button class="btn" data-job="${j.id}">Save site</button><button class="btn" data-now="${j.id}">Sync now (queue)</button><button class="btn" data-review="${j.clinic_id}">Review changes</button></div></details>`).join('')||'<p>Import/link a clinic from Syncro or UniFi first.</p>'}`;
    const status=card.querySelector('[role=status]');
    async function run(button,fn){button.disabled=true;try{await fn();await load();}catch(e){status.textContent=e.message;button.disabled=false;}}
    card.querySelector('[data-save]').onclick=e=>run(e.target,()=>api.put('/api/integration-sync/settings',{enabled:card.querySelector('[data-enabled]').checked}));
    card.querySelector('[data-reload]').onclick=()=>load().catch(e=>status.textContent=e.message);
    card.querySelectorAll('[data-job]').forEach(b=>b.onclick=()=>run(b,()=>api.put('/api/integration-sync/jobs/'+b.dataset.job,{enabled:card.querySelector(`[data-active="${b.dataset.job}"]`).checked,interval_minutes:Number(card.querySelector(`[data-interval="${b.dataset.job}"]`).value)})));
    card.querySelectorAll('[data-now]').forEach(b=>b.onclick=()=>run(b,()=>api.post(`/api/integration-sync/jobs/${b.dataset.now}/now`,{})));
    card.querySelectorAll('[data-review]').forEach(b=>b.onclick=()=>reviewChanges(Number(b.dataset.review)));
  }
  await load();
}

export async function syncClinic(container,cid){
  const user=await api.get('/api/auth/me');if(!user.roles.includes('admin')&&user.active_role!=='it')return;
  const data=await api.get('/api/integration-sync/clinics/'+cid);if(!data.jobs.length||!container.isConnected)return;
  const card=document.createElement('section');card.className='card mb';
  card.innerHTML=`<h2>Integration freshness</h2>${data.jobs.map(j=>`<p><strong>${esc(j.provider)} · ${esc(j.site_name)}</strong>: ${esc(state(j,data.enabled))}<br>Last successful sync: ${date(j.last_success)} · Next attempt: ${date(j.next_at)}<br>${j.pending} pending changes ${j.error?'· '+esc(j.error):''}</p>`).join('')}<button class="btn" data-review>Changes to review</button><p>“Not reported” does not mean offline or deleted. Source snapshots are dated observations.</p>`;
  container.append(card);card.querySelector('[data-review]').onclick=()=>reviewChanges(cid);
}

export async function reviewChanges(cid){
  const modal=openModal({title:'Integration changes to review',size:'modal-lg',body:'Loading…',footer:'<button class="btn" data-close>Close</button>'});
  modal.root.querySelector('[data-close]').onclick=()=>modal.close();
  async function load(){
    const [data,user]=await Promise.all([api.get('/api/integration-sync/clinics/'+cid),api.get('/api/auth/me')]);
    const admin=user.roles.includes('admin');
    modal.body.innerHTML=`<p>These observations have not changed your wiring or added/deleted machines. Use a reviewed import to add records, or edit the topology after checking the change. “Mark reviewed” acknowledges the observation only.</p>${admin?'<div class="actions"><button class="btn" data-syncro>Open Syncro import</button><button class="btn" data-unifi>Open UniFi import</button></div>':''}
      ${data.changes.map(c=>`<details class="card"><summary>${esc(c.title)} · ${date(c.updated_at)}</summary><pre style="white-space:pre-wrap">${esc(JSON.stringify(c.detail,null,2))}</pre>${admin?`<button class="btn" data-ack="${c.id}">Mark reviewed (no data changes)</button>`:''}</details>`).join('')||'<p>No pending changes.</p>'}`;
    modal.body.querySelectorAll('[data-ack]').forEach(b=>b.onclick=async()=>{b.disabled=true;try{await api.post('/api/integration-sync/changes/'+b.dataset.ack+'/acknowledge',{});await load();}catch(e){toast(e.message,'error');b.disabled=false;}});
    if(admin){
      modal.body.querySelector('[data-syncro]').onclick=async()=>{modal.close();(await import('./syncro.js')).importSyncro();};
      modal.body.querySelector('[data-unifi]').onclick=async()=>{modal.close();(await import('./unifi.js')).importUnifi(cid);};
      const meraki=document.createElement('button');meraki.className='btn';meraki.textContent='Open Meraki import';
      meraki.onclick=async()=>{modal.close();(await import('./meraki.js')).importMeraki(cid);};
      modal.body.querySelector('.actions').append(meraki);
    }
  }
  try{await load();}catch(e){modal.body.textContent=e.message;}
}
